src/renate/utils/pytorch.py

```diff
@@ -156,3 +156,76 @@ def complementary_indices(num_outputs: int, valid_classes: Set[int]) -> List[int
+    Args:
+        dataset_lengths: The length for the different datasets.
+        batch_sizes: Batch sizes used for specific datasets.
+        complete_dataset_iteration: Provide an index to indicate over which dataset to fully
+            iterate. By default, stops whenever iteration is complete for any dataset.
+        generator: Generator used in sampling.
+        sampler: Lightning automatically passes a DistributedSamplerWrapper. Only used as an
+            indicator that we are in the distributed case.
+    """
+
+    def __init__(
+        self,
+        dataset_lengths: List[int],
+        batch_sizes: List[int],
+        complete_dataset_iteration: Optional[int] = None,
+        generator: Any = None,
+        sampler: Sampler = None,
+    ) -> None:
+        self.batch_sizes = batch_sizes
+        self.complete_dataset_iteration = complete_dataset_iteration
+        self.subset_samplers = []
+        data_start_idx = 0
+        num_batches = []
+        rank = 0 if sampler is None else sampler.rank
+        num_replicas = 1 if sampler is None else sampler.num_replicas
+        for dataset_length, batch_size in zip(dataset_lengths, batch_sizes):
+            data_end_idx = data_start_idx + dataset_length
+            start_idx = data_start_idx + round(dataset_length / num_replicas * rank)
+            end_idx = data_start_idx + round(dataset_length / num_replicas * (rank + 1))
+            subset_sampler = BatchSampler(
+                SubsetRandomSampler(list(range(start_idx, end_idx)), generator),
+                batch_size,
+                True,
+            )
+            self.subset_samplers.append(subset_sampler)
+            num_batches.append((end_idx - start_idx) // batch_size)
+            data_start_idx = data_end_idx
+        self.length = (
+            min(num_batches)
+            if complete_dataset_iteration is None
+            else num_batches[self.complete_dataset_iteration]
+        )
+
+    def __iter__(self) -> Iterator[List[int]]:
+        """Creates a batch with groups of indices where each group corresponds to one dataset."""
+        if self.complete_dataset_iteration is None:
+            # Default case is iterating once over the shortest iterator. Works nicely with zip.
+            for samples in zip(*self.subset_samplers):
+                yield [j for i in samples for j in i]
+        else:
+            # Iterating over a specific iterator requires dealing with the length of other
+            # iterators.
+            iterators = [iter(sampler) for sampler in self.subset_samplers]
+            for s in iterators[self.complete_dataset_iteration]:
+                samples = []
+                for i, iterator in enumerate(iterators):
+                    if i != self.complete_dataset_iteration:
+                        try:
+                            samples += next(iterator)
+                        except StopIteration:
+                            iterators[i] = iter(self.subset_samplers[i])
```

Comment:
Can you add comments for this? 

Also, why not https://docs.python.org/3/library/itertools.html#itertools.zip_longest ?

Reply:
zip_longest requires a `fill_value`. however, we don't want to use a fixed `fill_value` but rather iterate over the other data sources.